### ISSUE TITLE
Add dependency for kubernetes_config_map on AKS 

### ISSUE THREAD

**Issue by manoatsas** [sassoftware/viya4-iac-azure#106]
When an error occurs on terraform (re)apply and destroy get this error - 
`Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo": dial tcp [::1]:80: connect: connection refused`

Any kubernetes resources should have dependency on K8s cluster resource and cleaned up accordingly.

**Comment by thpang** [comment]
I'll look into this. Have not seen this error in our current testing. Thx.

**Comment by manoatsas** [comment]
Yea it probably does not occur in a happy-path flow but when a re-apply fails, it where I hit this.
It seems to leave the resource in limbo and any further terraform commands are stopped in the tracks

```
terraform state list                                                      
data.azurerm_public_ip.aks_public_ip
data.azurerm_subnet.aks-subnet
data.azurerm_subnet.misc-subnet
data.azurerm_subscription.current
data.external.git_hash
data.external.iac_tooling_version
data.template_cloudinit_config.jump
data.template_cloudinit_config.nfs
data.template_file.jump-cloudconfig
data.template_file.nfs-cloudconfig
azurerm_network_security_group.nsg
azurerm_resource_group.azure_rg
kubernetes_config_map.sas_iac_buildinfo
module.jump.azurerm_public_ip.vm_ip[0]
module.nfs.azurerm_linux_virtual_machine.vm[0]
module.nfs.azurerm_managed_disk.vm_data_disk[0]
module.nfs.azurerm_managed_disk.vm_data_disk[1]
module.nfs.azurerm_managed_disk.vm_data_disk[2]
module.nfs.azurerm_managed_disk.vm_data_disk[3]
module.nfs.azurerm_network_interface.vm_nic[0]
module.nfs.azurerm_network_interface_security_group_association.vm_nic_sg[0]
module.nfs.azurerm_virtual_machine_data_disk_attachment.vm_data_disk_attach[0]
module.nfs.azurerm_virtual_machine_data_disk_attachment.vm_data_disk_attach[1]
module.nfs.azurerm_virtual_machine_data_disk_attachment.vm_data_disk_attach[2]
module.nfs.azurerm_virtual_machine_data_disk_attachment.vm_data_disk_attach[3]
module.vnet.data.azurerm_resource_group.vnet
module.vnet.azurerm_subnet.subnet[0]
module.vnet.azurerm_subnet.subnet[1]
module.vnet.azurerm_virtual_network.vnet
```
```
terraform destroy -target kubernetes_config_map.sas_iac_buildinfo  
azurerm_resource_group.azure_rg: Refreshing state... [id=/subscriptions/85704435-0cf9-4366-bf03-ef93f952145a/resourceGroups/mnm-tst-rg]
module.vnet.azurerm_virtual_network.vnet: Refreshing state... [id=/subscriptions/85704435-0cf9-4366-bf03-ef93f952145a/resourceGroups/mnm-tst-rg/providers/Microsoft.Network/virtualNetworks/mnm-tst-vnet]
module.vnet.azurerm_subnet.subnet[0]: Refreshing state... [id=/subscriptions/85704435-0cf9-4366-bf03-ef93f952145a/resourceGroups/mnm-tst-rg/providers/Microsoft.Network/virtualNetworks/mnm-tst-vnet/subnets/mnm-tst-aks-subnet]
module.vnet.azurerm_subnet.subnet[1]: Refreshing state... [id=/subscriptions/85704435-0cf9-4366-bf03-ef93f952145a/resourceGroups/mnm-tst-rg/providers/Microsoft.Network/virtualNetworks/mnm-tst-vnet/subnets/mnm-tst-misc-subnet]
kubernetes_config_map.sas_iac_buildinfo: Refreshing state... [id=kube-system/sas-iac-buildinfo]

Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo": dial tcp [::1]:80: connect: connection refused
```

**Comment by thpang** [comment]
Ok, I'll try the `depends_on` directive for the configMap and relate that to the aks module. I'll code something up and test it.

**Comment by manoatsas** [comment]
Adding dependency alone doesn't seem to address, as the issue still occurs. Debugging further and looking at a possible fix
```2021/02/22 11:34:52 [WARN] ReferenceTransformer: reference not found: "module.aks"
2021/02/22 11:34:52 [WARN] ReferenceTransformer: reference not found: "module.node_pools"
2021/02/22 11:34:52 [DEBUG] ReferenceTransformer: "data.azurerm_public_ip.aks_public_ip" references: []
2021-02-22T11:34:52.090-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 2021/02/22 11:34:52 [WARN] Invalid provider configuration was supplied. Provider operations likely to fail: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
2021-02-22T11:34:52.090-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 2021/02/22 11:34:52 [DEBUG] Enabling HTTP requests/responses tracing
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 2021/02/22 11:34:52 [INFO] Checking config map sas-iac-buildinfo
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 2021/02/22 11:34:52 [DEBUG] Kubernetes API Request Details:
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: ---[ REQUEST ]---------------------------------------
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: GET /api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo HTTP/1.1
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: Host: localhost
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: User-Agent: HashiCorp/1.0 Terraform/0.13.5
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: Accept: application/json, */*
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: Accept-Encoding: gzip
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 
2021-02-22T11:34:52.093-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: -----------------------------------------------------
2021-02-22T11:34:52.094-0500 [DEBUG] plugin.terraform-provider-kubernetes_v1.13.3_x4: 2021/02/22 11:34:52 [DEBUG] Received error: &url.Error{Op:"Get", URL:"http://localhost/api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo", Err:(*net.OpError)(0xc00013acd0)}
2021/02/22 11:34:52 [ERROR] eval: *terraform.EvalRefresh, err: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo": dial tcp [::1]:80: connect: connection refused
2021/02/22 11:34:52 [ERROR] eval: *terraform.EvalSequence, err: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/sas-iac-buildinfo": dial tcp [::1]:80: connect: connection refused```

**Comment by thpang** [comment]
What command did you run to generate the error above?

**Comment by enderm** [comment]
One way to replicate is to delete the k8s cluster with the cli or the portal and then run terraform destroy.

**Comment by manoatsas** [comment]
> What command did you run to generate the error above?

terraform plan, after updating tfvar values that force AKS to be recreated.

**Comment by thpang** [comment]
Ok, so if someone alters the infrastructure outside of terraform, this is really not an issue. If they are deleting items inside the UI, they should remove the terraform.tfstate files.

If this is a requirement that terraform needs to keep itself in sync when a user alters the infrastructure on the provider via the UI, this needs to be added so we can address.

**Comment by manoatsas** [comment]
Resolved with changes in PR #123
